File: services/identity-service/src/main/java/br/com/deltaglobalbank/identity/domain/token/RefreshToken.java
package br.com.deltaglobalbank.identity.domain.token;

import java.time.Instant;
import java.util.Objects;
import java.util.UUID;

public class RefreshToken {

    private final UUID id;
    private final UUID userId;
    private final String tokenHash;
    private final Instant expiresAt;
    private final Instant createdAt;
    private final String userAgent;
    private final String ipAddress;

    private Instant revokedAt;
    private Instant lastUsedAt;

    public RefreshToken(
        UUID id,
        UUID userId,
        String tokenHash,
        Instant expiresAt,
        Instant revokedAt,
        Instant createdAt,
        Instant lastUsedAt,
        String userAgent,
        String ipAddress
    ) {
        this.id = id;
        this.userId = userId;
        this.tokenHash = tokenHash;
        this.expiresAt = expiresAt;
        this.revokedAt = revokedAt;
        this.createdAt = createdAt;
        this.lastUsedAt = lastUsedAt;
        this.userAgent = userAgent;
        this.ipAddress = ipAddress;
    }

    public static RefreshToken newToken(
        UUID id,
        UUID userId,
        String tokenHash,
        Instant expiresAt,
        String userAgent,
        String ipAddress
    ) {
        return new RefreshToken(
            id,
            userId,
            tokenHash,
            expiresAt,
            null,
            Instant.now(),
            null,
            userAgent,
            ipAddress
        );
    }

    public UUID getId() {
        return id;
    }

    public UUID getUserId() {
        return userId;
    }

    public String getTokenHash() {
        return tokenHash;
    }

    public Instant getExpiresAt() {
        return expiresAt;
    }

    public Instant getCreatedAt() {
        return createdAt;
    }

    public String getUserAgent() {
        return userAgent;
    }

    public String getIpAddress() {
        return ipAddress;
    }

    public void revoke() {
        revokedAt = Instant.now();
    }

    public void markUsed() {
        lastUsedAt = Instant.now();
    }

    public boolean isActive() {
        return revokedAt == null && !isExpired();
    }

    public boolean isExpired() {
        return Instant.now().isAfter(expiresAt);
    }

    public boolean isRevoked() {
        return revokedAt != null;
    }

    public RefreshTokenSnapshot snapshot() {
        return new RefreshTokenSnapshot(
            id,
            userId,
            tokenHash,
            expiresAt,
            revokedAt,
            createdAt,
            lastUsedAt,
            userAgent,
            ipAddress
        );
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) {
            return true;
        }
        return other instanceof RefreshToken refreshToken && id.equals(refreshToken.id);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(id);
    }

    @Override
    public String toString() {
        return "RefreshToken(id=" + id + ", userId=" + userId + ")";
    }
}

File: services/identity-service/src/main/kotlin/br/com/deltaglobalbank/identity/domain/token/RefreshToken.kt
package br.com.deltaglobalbank.identity.domain.token

import java.time.Instant
import java.util.UUID

class RefreshToken(
    val id: UUID,
    val userId: UUID,
    val tokenHash: String,
    val expiresAt: Instant,
    revokedAt: Instant?,
    val createdAt: Instant,
    lastUsedAt: Instant?,
    val userAgent: String?,
    val ipAddress: String?
) {
    companion object {
        fun newToken(
            id: UUID,
            userId: UUID,
            tokenHash: String,
            expiresAt: Instant,
            userAgent: String?,
            ipAddress: String?
        ): RefreshToken {
            return RefreshToken(
                id = id,
                userId = userId,
                tokenHash = tokenHash,
                expiresAt = expiresAt,
                revokedAt = null,
                createdAt = Instant.now(),
                lastUsedAt = null,
                userAgent = userAgent,
                ipAddress = ipAddress
            )
        }
    }

    private var _revokedAt: Instant? = revokedAt
    private var _lastUsedAt: Instant? = lastUsedAt

    fun revoke() {
        _revokedAt = Instant.now()
    }

    fun markUsed() {
        _lastUsedAt = Instant.now()
    }

    fun isActive(): Boolean = _revokedAt == null && !isExpired()

    fun isExpired(): Boolean = Instant.now().isAfter(expiresAt)

    fun isRevoked(): Boolean = _revokedAt != null

    fun snapshot(): RefreshTokenSnapshot = RefreshTokenSnapshot(
        id = id,
        userId = userId,
        tokenHash = tokenHash,
        expiresAt = expiresAt,
        revokedAt = _revokedAt,
        createdAt = createdAt,
        lastUsedAt = _lastUsedAt,
        userAgent = userAgent,
        ipAddress = ipAddress
    )

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is RefreshToken) return false
        return id == other.id
    }

    override fun hashCode(): Int = id.hashCode()

    override fun toString(): String = "RefreshToken(id=$id, userId=$userId)"
}

data class RefreshTokenSnapshot(
    val id: UUID,
    val userId: UUID,
    val tokenHash: String,
    val expiresAt: Instant,
    val revokedAt: Instant?,
    val createdAt: Instant,
    val lastUsedAt: Instant?,
    val userAgent: String?,
    val ipAddress: String?
)
